### ISSUE TITLE
Creating per session cloud table on read requests

### ISSUE THREAD

**Issue by ozcelgozde** [ByConity/ByConity#1232]
## Question
I was looking into some timeouts during reads and noticed we are creating a cloud table over and over again for each read transaction even though table has not changed at all. This causes 3-10 seconds delay for each query and sometimes timeouts. I was wondering why read replicas dont keep a copy of the queried tables once its loaded until an modify event occurs in which case, they can reload?

**Comment by yuanzhubi** [comment]
Do you mean the StorageCloudMergeTree object in the worker? Different sub-query from server may search for different parts in the same table.(addDataParts, loadDataParts). So it should not be used outside the query session.

**Comment by ozcelgozde** [comment]
Yes I mean StorageCloudMergeTree object in the worker. I understand loading different parts, but recreating table every time creates an interesting error on retry or fallback when a timeout or rpc socket connection issue occurs:

2024.02.29 07:15:03.325072 [ 1494590 ] {3ba693d9-d6d2-11ee-892c-ba3d820a74a5} (448054137376211845) <Error> TCPHandler: Code: 57, e.displayText() = DB::Exception: DB::Exception: Table ed_mt_v1.ed_metric_2d6be233-f7bb-4fe1-90a5-28a95c86ec9c_448054137376211845 already exists. SQLSTATE: 42P07.: While executing Remote SQLSTATE: 42P07, Stack trace:
0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x223c0252 in /opt/byconity/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xebc1580 in /opt/byconity/bin/clickhouse
2. DB::readException(DB::ReadBuffer&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) @ 0xec30dcf in /opt/byconity/bin/clickhouse
3. DB::RPCHelpers::checkException(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x1b706f8b in /opt/byconity/bin/clickhouse
4. void DB::RPCHelpers::onAsyncCallDoneWithFailedInfo<DB::Protos::SendResourcesResp>(DB::Protos::SendResourcesResp*, brpc::Controller*, std::__1::shared_ptr<DB::ExceptionHandlerWithFailedInfo>, DB::WorkerId) @ 0x1b51522a in /opt/byconity/bin/clickhouse
5. brpc::internal::FunctionClosure4<DB::Protos::SendResourcesResp*, brpc::Controller*, std::__1::shared_ptr<DB::ExceptionHandlerWithFailedInfo>, DB::WorkerId>::Run() @ 0x1b516fa1 in /opt/byconity/bin/clickhouse
6. /home/ByConity/build_docker/../contrib/incubator-brpc/src/brpc/controller.h:704: brpc::Controller::EndRPC(brpc::Controller::CompletionInfo const&) @ 0x1e1ae9a9 in /opt/byconity/bin/clickhouse
7. /home/ByConity/build_docker/../contrib/incubator-brpc/src/brpc/controller.cpp:0: brpc::Controller::OnVersionedRPCReturned(brpc::Controller::CompletionInfo const&, bool, int) @ 0x1e1ad506 in /opt/byconity/bin/clickhouse
8. /home/ByConity/build_docker/../contrib/incubator-brpc/src/brpc/details/controller_private_accessor.h:0: brpc::policy::ProcessRpcResponse(brpc::InputMessageBase*) @ 0x1e1d6a30 in /opt/byconity/bin/clickhouse
9. /home/ByConity/build_docker/../contrib/incubator-brpc/src/brpc/input_messenger.cpp:386: brpc::InputMessenger::OnNewMessages(brpc::Socket*) @ 0x1e1cd2ce in /opt/byconity/bin/clickhouse
10. /home/ByConity/build_docker/../contrib/libcxx/include/memory:1655: brpc::Socket::ProcessEvent(void*) @ 0x1e2fd6cd in /opt/byconity/bin/clickhouse
11. /home/ByConity/build_docker/../contrib/incubator-brpc/src/bthread/task_group.cpp:304: bthread::TaskGroup::task_runner(long) @ 0x1e19af2d in /opt/byconity/bin/clickhouse

**Comment by yuanzhubi** [comment]
@smmsmm1988 Could you check about the exception and confirm whether it is a knowable issue?

**Comment by smmsmm1988** [comment]
> @smmsmm1988 Could you check about the exception and confirm whether it is a knowable issue?

No, it seems to be a new issue. @luffyhwl Could you help check this exception?

**Comment by smmsmm1988** [comment]
> > @smmsmm1988 Could you check about the exception and confirm whether it is a knowable issue?
> 
> No, it seems to be a new issue. @luffyhwl Could you help check this exception?

**Comment by luffyhwl** [comment]
> Yes I mean StorageCloudMergeTree object in the worker. I understand loading different parts, but recreating table every time creates an interesting error on retry or fallback when a timeout or rpc socket connection issue occurs:
> 
> 2024.02.29 07:15:03.325072 [ 1494590 ] {3ba693d9-d6d2-11ee-892c-ba3d820a74a5} (448054137376211845) TCPHandler: Code: 57, e.displayText() = DB::Exception: DB::Exception: Table ed_mt_v1.ed_metric_2d6be233-f7bb-4fe1-90a5-28a95c86ec9c_448054137376211845 already exists. SQLSTATE: 42P07.: While executing Remote SQLSTATE: 42P07, Stack trace: 0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits, std::__1::allocator > const&, int) @ 0x223c0252 in /opt/byconity/bin/clickhouse
> 
> 1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits, std::__1::allocator > const&, int, bool) @ 0xebc1580 in /opt/byconity/bin/clickhouse
> 2. DB::readException(DB::ReadBuffer&, std::__1::basic_string<char, std::__1::char_traits, std::__1::allocator > const&, bool) @ 0xec30dcf in /opt/byconity/bin/clickhouse
> 3. DB::RPCHelpers::checkException(std::__1::basic_string<char, std::__1::char_traits, std::__1::allocator > const&) @ 0x1b706f8b in /opt/byconity/bin/clickhouse
> 4. void DB::RPCHelpers::onAsyncCallDoneWithFailedInfoDB::Protos::SendResourcesResp(DB::Protos::SendResourcesResp*, brpc::Controller*, std::__1::shared_ptrDB::ExceptionHandlerWithFailedInfo, DB::WorkerId) @ 0x1b51522a in /opt/byconity/bin/clickhouse
> 5. brpc::internal::FunctionClosure4<DB::Protos::SendResourcesResp*, brpc::Controller*, std::__1::shared_ptrDB::ExceptionHandlerWithFailedInfo, DB::WorkerId>::Run() @ 0x1b516fa1 in /opt/byconity/bin/clickhouse
> 6. /home/ByConity/build_docker/../contrib/incubator-brpc/src/brpc/controller.h:704: brpc::Controller::EndRPC(brpc::Controller::CompletionInfo const&) @ 0x1e1ae9a9 in /opt/byconity/bin/clickhouse
> 7. /home/ByConity/build_docker/../contrib/incubator-brpc/src/brpc/controller.cpp:0: brpc::Controller::OnVersionedRPCReturned(brpc::Controller::CompletionInfo const&, bool, int) @ 0x1e1ad506 in /opt/byconity/bin/clickhouse
> 8. /home/ByConity/build_docker/../contrib/incubator-brpc/src/brpc/details/controller_private_accessor.h:0: brpc::policy::ProcessRpcResponse(brpc::InputMessageBase*) @ 0x1e1d6a30 in /opt/byconity/bin/clickhouse
> 9. /home/ByConity/build_docker/../contrib/incubator-brpc/src/brpc/input_messenger.cpp:386: brpc::InputMessenger::OnNewMessages(brpc::Socket*) @ 0x1e1cd2ce in /opt/byconity/bin/clickhouse
> 10. /home/ByConity/build_docker/../contrib/libcxx/include/memory:1655: brpc::Socket::ProcessEvent(void*) @ 0x1e2fd6cd in /opt/byconity/bin/clickhouse
> 11. /home/ByConity/build_docker/../contrib/incubator-brpc/src/bthread/task_group.cpp:304: bthread::TaskGroup::task_runner(long) @ 0x1e19af2d in /opt/byconity/bin/clickhouse

@ozcelgozde 
This exception may occur during fallback. The commit below fixes this issue. When sql fails to execute in optimizer mode,  worker resource will be synchronized delete.

commit ac18e6a238905b761af723f6d56d89bc9ec83fc0
Author: Dao <minh.dao@bytedance.com>
Date:   Thu Jan 18 10:32:52 2024 +0800

    Merge branch 'zema/cnch-2.0-fixRemoveWorkerResource' into 'cnch-ce-merge'
    
    fix(clickhousech@m-3000501254): remove SUBMIT_THREADPOOL for removeWorkerResource

**Comment by ozcelgozde** [comment]
Thanks for the quick fix! Do we have an idea why optimizer fails noticeably often?

**Comment by ozcelgozde** [comment]
Any number of possible errors return table already exists :)
